icu4c/source/common/locid.cpp

```diff
@@ -633,8 +633,21 @@ class AliasDataBuilder {
  */
 class AliasData : public UMemory {
 public:
-    static const AliasData* singleton(UErrorCode status) {
+    static const AliasData* singleton(UErrorCode& status) {
+        if (U_FAILURE(status)) {
+            // Do not get into loadData if the status already has error.
+            return nullptr;
+        }
         umtx_initOnce(AliasData::gInitOnce, &AliasData::loadData, status);
+        if (U_FAILURE(status)) {
```

Comment:
optional: This works, but if loadData() returns nullptr if-and-only-if U_SUCCESS(status) then this singleton() function should be able to return gSingleton without doing anything else.

Reply:
the first time the singleton() got called, it will cause AliasData::loadData to be called, and AliasData::loadData may set the status to error and gSingleton to nullptr. if we do not check the error and return here, then the next line will clobber the error to U_MEMORY_ALLOCATION_ERROR but the error might be some other error.

Reply:
What I mean is, after your latest commit (thanks!), where the build() function becomes well-behaved, you should be able to shorten this function here to just:
```
    static const AliasData* singleton(UErrorCode& status) {
        if (U_FAILURE(status)) {
            // Do not get into loadData if the status already has error.
            return nullptr;
        }
        umtx_initOnce(AliasData::gInitOnce, &AliasData::loadData, status);
        return gSingleton;
    }
```

Reply:
I don't understand how could I do this.
let's say this got called twice, the first time it come in, umtx_initOnce called AliasData::loadData and caused error, the status is set to error and gSingleton is set to nullptr
now the second time it is called, the umtx_initOnce skip calling AliasData::loadData because it has been called once, the status is still no error and gSingleton is still nullptr
and we return gSingleton, which mean we return nullptr

Then... in AliasReplacer::replace, I check the status and return if it is error, but the status is not error because singleton neither call  AliasData::loadData nor set the error as what you suggested above, so the code continue, and the next line is U_ASSERT(data != nullptr)
then it will bark if we run under a debug build but will continue if it is on a release build and then hit the same problem we have before.

Reply:
ok, I just realize umtx_initOnce would cache the status from the first load and set it to the later calls even it skip the calling for loadData. Now it make sense to me.

Reply:
The failure code from the first load is cached in the UInitOnce object and umtx_initOnce() will set it again on subsequent calls. You need not do anything special to remember the failure.